### PR TITLE
Fix LYS tour is shown again after closing it

### LINKED ISSUE #49126: LYS tour is shown again after closing it
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

After closing the LYS tour, it was shown again if a full reload was not performed.  

https://github.com/woocommerce/woocommerce/assets/4344253/65401768-20f0-41e4-a8ff-0e7cdd4913a2

### Expected behavior

The tour should not shown again after closing it.

### Actual behavior

The tour is shown again after closing it.

### Steps to reproduce

1. Create a fresh JN site
2. Set `woocommerce_show_lys_tour` option to `yes`.
3. Go to WooCommerce > Home
4. Confirm the tour is shown
5. Click on the product task
6. Click on the back icon
7. Observe that the tour is shown again

### WordPress Environment

N/A

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by chihsuan**
a### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #49126.

Invalidating the tour state after closing it, so it won't be shown again.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site with this branch
2. Set `woocommerce_show_lys_tour` option to yes.
3. Go to `WooCommerce > Home`
4. Confirm the tour is shown and close it
5. Click on the product task
6. Click on the back icon
7. Confirm the tour is not shown again

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx [hunk 1]
```diff
@@ -31,15 +31,18 @@ export const useSiteVisibilityTour = () => {
...
 
-	const onClose = () => {
+	const onClose = async () => {
 		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
 		// @ts-ignore
-		dispatch( 'core' ).saveUser( {
+		await dispatch( 'core' ).saveUser( {
 			id: window?.wcSettings?.currentUserId,
 			meta: {
 				woocommerce_launch_your_store_tour_hidden: 'yes',
```
Comment (psealock) [inline plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx]: There's a `const` for this key above, may as well use it here too.

File: plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx [hunk 2]
```diff
@@ -31,15 +31,18 @@ export const useSiteVisibilityTour = () => {
...
 			},
 		} );
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		dispatch( 'core' ).invalidateResolution( 'getCurrentUser' );
```
Comment (psealock) [inline plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx:45]: Interesting that you'd need to invalidate the last resolution here. Any idea why `shouldTourBeShown` from the `useSelect` isn't reading the latest data in its `getCurrentUser` call?

**Review by psealock — COMMENTED** [review]
Thanks @chihsuan, this approach seems heavy handed I wouldn't think you'd need to do this. Does [useUserPreferences](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/user/use-user-preferences.ts) help here at all?

File: plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx]: Sounds good! Updated in e8047cfe2d294df02065eab9e880379de3785a0e

File: plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx (hunk 2 above)
Comment (chihsuan) [reply plugins/woocommerce-admin/client/launch-your-store/tour/use-site-visibility-tour.tsx:45]: I think it's related to how the data flow and cache management work in wp core data. 🤔  The `saveUser` call doesn't update the local currentUser state after the user is saved to the server. So, the `shouldTourBeShown` function is reading the old data from the cache.

https://github.com/user-attachments/assets/b49b0bfc-342f-4ae9-b8a5-dd24010f69d1

Screenshot 2024-08-13 at 16 25 31

**Comment by chihsuan** [comment]
> Does [useUserPreferences](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/user/use-user-preferences.ts) help here at all?

@psealock Unfortunately, no. `useUserPreferences` is used to get the user preferences from the `woocommerce_meta` attribute, but we store the `woocommerce_launch_your_store_tour_hidden` in `meta` attribute ( It should have been stored in `woocommerce_meta`  🥲 ).

**Comment by adrianduffell** [comment]
> useUserPreferences is used to get the user preferences from the woocommerce_meta attribute, but we store the woocommerce_launch_your_store_tour_hidden in meta attribute 

@chihsuan @psealock Ah right 😅  I think we need to tidy this up in a db update. Perhaps we should do that first before this PR?

**Comment by chihsuan** [comment]
> @chihsuan @psealock Ah right 😅 I think we need to tidy this up in a db update. Perhaps we should do that first before this PR?

@adrianduffell Yes, if we want to clean this up, we should start with that. However, I think it might require more effort.

Since users need to click the "Update Database" button to migrate their data, we must ensure that existing users don't see the tour again before clicking the button. So, we need to keep the existing checking logic in place to manage this situation.

Here is the plan to clean this up:

WC 9.3: Implement the database update and maintain the old and new logic.
WC 9.4: Remove the old logic.

Does that make sense?

**Comment by adrianduffell** [comment]
> Does that make sense?

@chihsuan Gotcha makes sense, thanks for clarifying the extra nuance.

I think it's still worth pursuing so we don't create tech debt, especially now while we have a good grasp of the situation.

~~However would we be better to take the heavy handed approach in the meantime so there is a fix for 9.3?~~

Edit:  I'm in favor of delaying the eventual fix to 9.4 if it keeps things simpler.

**Comment by adrianduffell** [comment]
I've also been working on updating user meta in a db update for this [issue](https://github.com/woocommerce/woocommerce/issues/43300). Here's the gist of what I've come up with:

```
function wc_update_930_remove_help_panel_user_meta() {

    // Query users who have 'woocommerce_meta' user meta
    $users_with_meta = get_users( array(
        'meta_key' => 'woocommerce_meta',
        'meta_compare' => 'EXISTS',
    ));

    // Loop through each user
    foreach ( $users_with_meta as $user ) {
        // Todo: update the meta here
    }
}
```

**Comment by chihsuan** [comment]
> I think it's still worth pursuing so we don't create tech debt, especially now while we have a good grasp of the situation.
> 
> ~However would we be better to take the heavy handed approach in the meantime so there is a fix for 9.3?~
> 
> Edit: I'm in favor of delaying the eventual fix to 9.4 if it keeps things simpler.

@adrianduffell Sounds good to me! I'll close this PR and work on a new one for the meta update.

> I've also been working on updating user meta in a db update for this https://github.com/woocommerce/woocommerce/issues/43300. Here's the gist of what I've come up with:

Thanks for sharing this!

**Comment by chihsuan** [comment]
@adrianduffell FYI, I found that we use the `woocommerce_admin_` prefix for woocommerce user meta keys.

https://github.com/woocommerce/woocommerce/blob/5c980cc6299af86ed18bb8cb10695364467db761/plugins/woocommerce/src/Internal/Admin/WCAdminUser.php#L127

So I can simply rename the meta key. 🙂 

````php
	global $wpdb;
	$wpdb->query(
		$wpdb->prepare(
			"UPDATE {$wpdb->usermeta}
			SET meta_key = %s
			WHERE meta_key = %s",
			'woocommerce_admin_launch_your_store_tour_hidden',
			'woocommerce_launch_your_store_tour_hidden'
		)
	);
```

**Comment by adrianduffell** [comment]
> So I can simply rename the meta key

@chihsuan Nice, that's looking good!

**Pull request closed without merging**